### ISSUE TITLE
LongTensor indexing with duplication not propagating gradients

### ISSUE THREAD

**Issue by zkolter** [pytorch/pytorch#1528]
It seems like if we use LongTensor indexing to duplicate entries of a Variable, this breaks autograd. Specifically, if any entry in a vector is duplicated, backward() only maintains gradients for the last copy of the element. Simple example:

```python
y = Variable(torch.ones(1), requires_grad=True)
y_dup = y[torch.LongTensor([0,0])]
y_dup[1].backward() # backprop on second element
print(y.grad)
"""
Variable containing:
 1
[torch.FloatTensor of size 1]"""
```

However:
```python
y = Variable(torch.ones(1), requires_grad=True)
y_dup = y[torch.LongTensor([0,0])]
y_dup[0].backward()  # backprop on first element, same exact value as second
print(y.grad)
"""
Variable containing:
 0
[torch.FloatTensor of size 1]"""
```

This is version under version 0.1.11+fc48d2c, tested on 0.1.11+9f3119a as well.  This happens for both cuda and cpu tensors.

**Comment by zkolter** [comment]
FYI, realized I had been testing the double backprop branch, but this also happens with 0.1.12_2.

**Comment by zkolter** [comment]
Less critical than I had originally though (still worth fixing, but there's a viable alternative), as it looks like `torch.index_select` has the right behavior, even if it's a little odd that it requires a Variable third argument:

```python
y = Variable(torch.ones(1), requires_grad=True)
y_dup = torch.index_select(y, 0, Variable(torch.LongTensor([0,0])))
y_dup[0].backward() # backprop on first 
print(y.grad)
"""
In [42]: Variable containing:
 1
[torch.FloatTensor of size 1]"""
```

**Comment by apaszke** [comment]
Yes, it's because backward of assignments does `grad[index] = grad_output`, and this uses `index_copy_` instead of `index_add_` internally. `index_select` should be correct

**Comment by Stonesjtu** [comment]
Additionally, I have another example for this bug.
```python
va = Variable(torch.Tensor([1,1]), requires_grad=True)
idx = torch.LongTensor([0,0,0])
loss = va[idx].sum()
loss.backward()
print(va.grad)
```
output:
```
Variable containing:
 1  # supposed to be 3
 0
[torch.FloatTensor of size 2]
```
It costs me 2 weeks to locate this bug in my script.
Plz fix ASAP.

**Comment by apaszke** [comment]
I'll push a fix today
